src/_realizeCustomElements.ts

```diff
@@ -0,0 +1,398 @@
+	element: Element;
+	name: string;
+	widget?: WidgetLike;
+}
+
+function isCustomElement(registry: CombinedRegistry, name: string): boolean {
+	return name === 'projection-surface' || name === 'attach-widget' || registry.hasCustomElementFactory(name);
+}
+
+function getCustomElementsByProjectionSurface(registry: CombinedRegistry, root: Element): CustomElement[] {
+	const allElements: Element[] = Array.prototype.slice.call(root.getElementsByTagName('*'));
+	allElements.unshift(root); // Be inclusive!
+
+	const customElements: CustomElement[] = [];
+	for (const element of allElements) {
+		let name: string;
+
+		const tagName = normalizeName(element.tagName);
+		if (isCustomElement(registry, tagName)) {
+			name = tagName;
+		}
+		else {
+			const attrIs = normalizeName(element.getAttribute('is') || '');
+			if (attrIs !== '' && isCustomElement(registry, attrIs)) {
+				name = attrIs;
+			}
+		}
+
+		if (name) {
+			customElements.push({ children: [], element, name });
+		}
+	}
+
+	// A list of trees, reconstructed from the `customElements`.
+	const surfaces: CustomElement[] = [];
+	// Inverse stack of the nodes in the current tree. The deepest node is at the start of the list.
+	const inverseStack: CustomElement[] = [];
+
+	const discardFirstNode = (element: Element) => {
+		if (inverseStack.length === 0) {
+			return false;
+		}
+
+		// Return `true` if the top-most element in the stack does *not* contain `element`.
+		return !(inverseStack[0].element.compareDocumentPosition(element) & Node.DOCUMENT_POSITION_CONTAINED_BY);
+	};
+
+	// `customElements` is a flat list of elements, in document order. Reconstruct a tree structure where each
+	// root is assumed to be a projection surface.
+	for (const custom of customElements) {
+		// Remove nodes from the stack that do not contain the element.
+		while (discardFirstNode(custom.element)) {
+			inverseStack.shift();
+		}
+
+		// Start a new tree if the element is not contained in any existing node.
+		if (inverseStack.length === 0) {
+			// Don't costruct an invalid tree.
+			if (custom.name !== 'projection-surface') {
+				throw new Error('Custom tags must be rooted in a projection-surface');
+			}
+
+			surfaces.push(custom);
+		}
+		// Add the element to the deepest node it is contained by.
+		else {
+			// Don't costruct an invalid tree.
+			if (custom.name === 'projection-surface') {
+				throw new Error('projection-surface cannot contain another projection-surface');
+			}
+
+			inverseStack[0].children.push(custom);
+		}
+
+		// Prepare for the next iteration.
+		inverseStack.unshift(custom);
+	}
+
+	return surfaces;
+}
+
+function getIdFromAttributes(element: Element): string {
+	return element.getAttribute('data-widget-id') || element.getAttribute('id') || null;
+}
+
+interface Options {
+	id?: string;
+	listeners?: any;
+	stateFrom?: any;
+}
+
+function resolveOptions(registry: CombinedRegistry, element: Element, idFromAttributes: string): void | Options | Promise<Options> {
```

Comment:
Yea. I think I put in `void` as a kind of placeholder for `null`, but there isn't much value in that.